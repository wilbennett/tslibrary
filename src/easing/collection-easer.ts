import { EaserCallback, EaserValueCallback, TypedEaser } from '.';

export abstract class CollectionEaser<T> extends TypedEaser<T> {
  constructor(
    protected _easers: TypedEaser<T>[],
    onValueChanged: EaserValueCallback<T> = () => { },
    onComplete: EaserCallback = () => { }
  ) {
    super(1, onComplete);

    this.onValueChanged = onValueChanged;

    let duration = _easers.reduce((sum, easer) => sum + easer.duration, 0);
    duration += (_easers.length - 1) * this._secondsPerFrame;
    this._duration = duration;
  }

  protected _easeIndex: number = 0;
  get easer() { return this._easers[this._easeIndex]; }
  get ease() { return this.easer.ease; }
  get value() { return this.easer.value; }
  get start() { return this.easer.start; }
  get end() { return this.easer.end; }

  reset() {
    super.reset();
    this._easers.forEach(easer => easer.reset());
    this._easeIndex = 0;
  }

  reverse() {
    super.reverse();
    this._easers.forEach(easer => easer.reverse());
    this._easers = this._easers.reverse();
  }

  protected _iterationsRun = 0;

  moveNext() { return super.moveNext(); }

  protected nextEaser() {
    if (this._easeIndex >= this._easers.length - 1) return;

    this._easeIndex++;
    this.easer.moveNext();
  }

  // @ts-ignore - unused param.
  protected calcValue(percent: number): T { return this.easer.value; }
}