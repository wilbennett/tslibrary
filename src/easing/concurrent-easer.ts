import { CollectionEaser, Easer, EaserCallback } from '.';

export class ConcurrentEaser extends CollectionEaser {
  constructor(
    easers: Easer[],
    onComplete?: EaserCallback
  ) {
    super(easers, onComplete);

    let duration = this._easers.reduce((max, easer) => Math.max(max, easer.duration), 0);
    this._duration = duration;
    this.init();
  }

  protected _pendingEasers: Easer[] = [];

  reset() {
    super.reset();
    this._pendingEasers = [...this._easers];
  }

  moveNext() {
    const pending = this._pendingEasers;

    for (let i = pending.length - 1; i >= 0; i--) {
      const easer = pending[i];

      if (!easer.moveNext())
        pending.remove(easer);
    }

    return super.moveNext();
  }
}
