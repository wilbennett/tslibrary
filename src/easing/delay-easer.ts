import { ValueEaser, EaserValueCallback, EaserCallback, Ease } from ".";

export class DelayEaser extends ValueEaser<number> {
    constructor(
        duration: number,
        onValueChanged?: EaserValueCallback<number>,
        onComplete?: EaserCallback
    ) {
        super(0, 1, duration, Ease.linear, onValueChanged, onComplete);

        this.init();
    }

    // @ts-ignore - unused param.
    protected calcValue(percent: number): number {
        return this._steps - this._stepsRemain - 1;
    }
}