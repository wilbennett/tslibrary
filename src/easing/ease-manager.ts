import { EaseFunction, EaseInfo, Ease } from ".";

export class EaseManager {
    private readonly _lookup: Map<string, EaseInfo> = new Map<string, EaseInfo>();

    private _infos: EaseInfo[] | null = null;
    protected get infos() {
        if (!this._infos) {
            this._infos = Array.from(this._lookup.values());
        }

        return this._infos;
    }

    private static _instance: EaseManager = new EaseManager();
    static get instance(): EaseManager { return this._instance; }
    static set instance(value: EaseManager) { this._instance = value; }

    add(info: EaseInfo): void;
    add(name: string, ease: EaseFunction): void;
    add(param1: EaseInfo | string, ease?: EaseFunction): void {
        let info: EaseInfo;

        if (param1 instanceof EaseInfo)
            info = param1;
        else
            info = new EaseInfo(param1, ease!);

        this._lookup.set(info.name, info);
        this._infos = null;
    }

    addFrom(obj: object) {
        for (const name of Object.getOwnPropertyNames(obj)) {
            if (name === "length" || name === "name" || name === "prototype" || name === "constructor")
                continue;

            // @ts-ignore - indexer.
            const prop = obj[name];

            if (!(typeof prop === "function")) continue;

            this.add(name, prop);
        }
    }

    get(name: string) { return this._lookup.get(name); }

    getRandom(): EaseInfo;
    getRandom(count: number): EaseInfo[];
    getRandom(count?: number): EaseInfo | EaseInfo[] {
        if (count === undefined) {
            const index = Math.floor(Math.random() * this._lookup.size);
            return this.infos[index];
        }

        const result = new Array<EaseInfo>(count);

        for (let i = 0; i < count; i++) {
            const index = Math.floor(Math.random() * this._lookup.size);
            result[i] = this.infos[index];
        }

        return result;
    }

    getRandomEase(): EaseFunction;
    getRandomEase(count: number): EaseFunction[];
    getRandomEase(count?: number): EaseFunction | EaseFunction[] {
        if (count === undefined) {
            const index = Math.floor(Math.random() * this._lookup.size);
            return this.infos[index].ease;
        }

        const result = new Array<EaseFunction>(count);

        for (let i = 0; i < count; i++) {
            const index = Math.floor(Math.random() * this._lookup.size);
            result[i] = this.infos[index].ease;
        }

        return result;
    }

    static add(info: EaseInfo): void;
    static add(name: string, ease: EaseFunction): void;
    // @ts-ignore - unused params.
    static add(param1: EaseInfo | string, ease?: EaseFunction): void {
        // @ts-ignore - rest params.
        this.instance.add(...arguments);
    }

    static addFrom(obj: object) { this.instance.addFrom(obj); }
    static get(name: string) { return this.instance.get(name); }

    static getRandom(): EaseInfo;
    static getRandom(count: number): EaseInfo[];
    // @ts-ignore - unused params.
    static getRandom(count?: number): EaseInfo | EaseInfo[] {
        // @ts-ignore - rest params.
        return this.instance.getRandom(...arguments);
    }

    static getRandomEase(): EaseFunction;
    static getRandomEase(count: number): EaseFunction[];
    // @ts-ignore - unused params.
    static getRandomEase(count?: number): EaseFunction | EaseFunction[] {
        // @ts-ignore - rest params.
        return this.instance.getRandomEase(...arguments);
    }
}

EaseManager.instance.addFrom(Ease);
