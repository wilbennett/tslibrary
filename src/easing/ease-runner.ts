import { Easer, EaserCallback } from '.';
import { AnimationLoop } from '../animation';

export class EaseRunner {
    private _easers: Easer[] = [];
    private _onRemoveListeners: EaserCallback[] = [];
    private _isLoopOwner = false;

    constructor(loop?: AnimationLoop) {
        if (!loop) {
            loop = new AnimationLoop();
            this._isLoopOwner = true;
        }

        this._loop = loop;
    }

    private static _instance: EaseRunner = new EaseRunner(new AnimationLoop());
    static get instance(): EaseRunner { return this._instance; }
    static set instance(value: EaseRunner) { this._instance = value; }

    isPaused = false;

    private _isActive = false;
    get isActive() { return this._isActive; }

    private _loop: AnimationLoop;
    get loop(): AnimationLoop { return this._loop; }
    set loop(value: AnimationLoop) {
        if (this._loop) {
            this._loop.removeOnUpdate(this.update);
        }

        this._loop = value;

        if (this.isActive)
            this._loop.onUpdate(this.update);
    }

    add(...easers: Easer[]) { this._easers.push(...easers); }

    remove(easer: Easer) {
        const index = this._easers.findIndex(x => x === easer);

        if (index < 0) return;

        this._easers.splice(index, 1);
        this._onRemoveListeners.forEach(listener => listener(easer));
    }

    start() {
        if (this._isActive) return;

        this._isActive = true;
        this._loop.onUpdate(this.update);

        if (!this._loop.active)
            this._loop.start();
    }

    stop() {
        if (!this._isActive) return;

        this._isActive = false;
        this._loop.removeOnUpdate(this.update);
        this.removeEasers(this._easers.slice());

        if (this._isLoopOwner)
            this._loop.stop();
    }

    onRemove(listener: EaserCallback) { this._onRemoveListeners.push(listener); }

    removeOnRemove(listener: EaserCallback) {
        const index = this._onRemoveListeners.findIndex(x => x === listener);

        if (index < 0) return;

        this._onRemoveListeners.splice(index, 1);
    }

    private update = () => {
        if (this.isPaused) return;

        const toRemove: Easer[] = [];

        this._easers.forEach(easer => {
            if (!easer.moveNext())
                toRemove.push(easer);
        });

        this.removeEasers(toRemove);
    }

    private removeEasers(easers: Easer[]) {
        easers.forEach(easer => this.remove(easer));
    }
}
