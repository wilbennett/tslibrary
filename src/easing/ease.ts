import { EaseFunction } from ".";

export class Ease {
    static linear(t: number) { return t; }
    // One nice thing about smooth/smoother step is you can apply
    // them to other eases.
    static smoothStep(t: number) { return t * t * (3 - 2 * t); }
    static smootherStep(t: number) { return t * t * t * (t * (t * 6 - 15) + 10); }
    static inQuad(t: number) { return t * t; }
    static inCubic(t: number) { return t * t * t; }
    static inQuartic(t: number) { return t * t * t * t; }
    static inBack(t: number) {
        const s = 1.70158;
        return t * t * ((s + 1) * t - s);
    }
    static inElastic(t: number) {
        return t === 0 ? 0 : t === 1 ? 1 : (0.04 - 0.04 / t) * Math.sin(25 * t) + 1;
    }
    static inCatmullRomN84(t: number) { return catmullRom(t, -8, 4); }
    static inCatmullRom10(t: number) { return catmullRom(t, 1, 0); }
    static inCatmullRom2N1(t: number) { return catmullRom(t, 2, -1); }
    static inCatmullRom010(t: number) { return catmullRom(t, 0, 10); }
    static inCatmullRomN101(t: number) { return catmullRom(t, -10, 1); }

    static outQuad(t: number) { t = 1 - t; return 1 - (t * t); }
    static outCubic(t: number) { t = 1 - t; return 1 - (t * t * t); }
    static outQuartic(t: number) { t = 1 - t; return 1 - (t * t * t * t); }
    static outBack(t: number) {
        const s = 1.70158;
        t = 1 - t;
        return 1 - (t * t * ((s + 1) * t - s));
    }
    static outElastic(t: number) { return 0.04 * t / (--t) * Math.sin(25 * t); }
    static outCatmullRomN84(t: number) { t = 1 - t; return 1 - catmullRom(t, -8, 4); }
    static outCatmullRom10(t: number) { t = 1 - t; return 1 - catmullRom(t, 1, 0); }
    static outCatmullRom2N1(t: number) { t = 1 - t; return 1 - catmullRom(t, 2, -1); }
    static outCatmullRom010(t: number) { t = 1 - t; return 1 - catmullRom(t, 0, 10); }
    static outCatmullRomN101(t: number) { t = 1 - t; return 1 - catmullRom(t, -10, 1); }


    static inOutQuad(t: number) { return inOut(Ease.inQuad, t); }
    static inOutCubic(t: number) { return inOut(Ease.inCubic, t); }
    static inOutQuartic(t: number) { return inOut(Ease.inQuartic, t); }
    static inOutBack(t: number) { return inOut(halfBack, t); }
    static inOutElastic(t: number) { return inOut(Ease.inElastic, t); }
    static inOutCatmullRomN84(t: number) { return inOut(Ease.outCatmullRomN84, t); }
    static inOutCatmullRom10(t: number) { return inOut(Ease.outCatmullRom10, t); }
    static inOutCatmullRom2N1(t: number) { return inOut(Ease.outCatmullRom2N1, t); }
    static inOutCatmullRom010(t: number) { return inOut(Ease.outCatmullRom010, t); }
    static inOutCatmullRomN101(t: number) { return inOut(Ease.outCatmullRomN101, t); }
}

function inOut(ease: EaseFunction, t: number) {
    return t < 0.5 ? ease(t * 2) * 0.5 : 1 - ease((1 - t) * 2) * 0.5;
}

function halfBack(t: number) {
    const s = 1.70158 * 1.525;
    return t * t * ((s + 1) * t - s);
}

function catmullRom(t: number, p0: number, p3: number) {
    const p1 = 0;
    const p2 = 1;

    return 0.5 * (
        (2 * p1)
        + (-p0 + p2) * t
        + (2 * p0 - 5 * p1 + 4 * p2 - p3) * t * t
        + (-p0 + 3 * p1 - 3 * p2 + p3) * t * t * t
    );
}
