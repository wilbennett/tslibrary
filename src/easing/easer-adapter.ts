import { TypedEaser } from '.';
import { Easer } from './easer';

export class EaserAdapter extends TypedEaser<any> {
  constructor(public readonly easer: Easer) {
    super(easer.duration);
  }

  get value() { return 0; }
  get start() { return 0; }
  get end() { return 0; }
  get ease() { return this.easer.ease; }

  reset() {
    super.reset();
    this.easer.reset();
  }

  reverse() {
    super.reverse();
    this.easer.reverse();
  }

  moveNext() {
    this.advanceEaser();
    return super.moveNext();
  }

  protected advanceEaser() { this.easer.moveNext(); }
  protected calcStepAmt() { return 1 / this.calcSteps(); }
  protected calcSteps() { return this.easer.steps; }

  // @ts-ignore - unused param.
  protected calcValue(percent: number) { return 0; }
}
