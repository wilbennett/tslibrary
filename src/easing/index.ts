export * from "./easing-types";
export * from "./ease";
export * from "./ease-manager";
export * from "./easer";
export * from "./typed-easer";
export * from "./value-easer";
export * from "./easer-decorator";
export * from "./collection-easer";
export * from "./number-easer";
export * from "./array-easer";
export * from "./vector-easer";
export * from "./delay-easer";
export * from "./ping-pong-easer";
export * from "./repeat-easer";
export * from "./sequential-easer";
export * from "./concurrent-easer";
export * from "./easer-extensions";
export * from "./ease-runner";
