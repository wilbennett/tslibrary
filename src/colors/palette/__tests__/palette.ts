import { Palette } from '..';
import { CanvasColor, Color, ColorUtils, WebColors } from '../..';
import { Ease } from '../../../easing';

describe.only("Should create palettes", () => {
    const paletteLength = 10;
    const refCanvasColors: CanvasColor[] = [WebColors.red, WebColors.green, "blue"];
    const refColors = ColorUtils.toColorArray(...refCanvasColors);
    const stops = [0, 0.5, 1];
    let colors: Color[] = [];

    function validate(paletteOrColors: Palette | Color[]): void {
        if (paletteOrColors instanceof Palette) {
            expect(paletteOrColors.colors).toEqual(colors);
            expect(paletteOrColors.colorStops).toEqual(stops);
        } else {
            expect(paletteOrColors).toEqual(colors);
        }
    }

    it("Should error with invalid parameters", () => {
        const stops1 = stops.slice(0, 0);
        const refCanvasColors1 = refCanvasColors.slice(0, 0);
        const refColors1 = refColors.slice(0, 0);

        expect(() => new Palette(2, stops1, ...refCanvasColors)).toThrow();
        expect(() => new Palette(2, Ease.linear, stops1, ...refCanvasColors)).toThrow();
        expect(() => new Palette(2, Ease.linear, ...refCanvasColors1)).toThrow();
        expect(() => new Palette(1, ...refCanvasColors)).toThrow();

        expect(() => new Palette("hsl", 2, stops1, ...refCanvasColors)).toThrow();
        expect(() => new Palette("hsl", 2, Ease.linear, stops1, ...refCanvasColors)).toThrow();
        expect(() => new Palette("hsl", 2, Ease.linear, ...refCanvasColors1)).toThrow();
        expect(() => new Palette("hsl", 1, ...refCanvasColors)).toThrow();

        expect(() => Palette.getColors(1, refColors, stops1, "rgb", Ease.linear, colors)).toThrow();
        expect(() => Palette.getColors(2, refColors, stops1, "rgb", Ease.linear, colors)).toThrow();
        expect(() => Palette.getColors(2, refColors1, stops, "rgb", Ease.linear, colors)).toThrow();
    });

    it("Should allow creating Palette instance", () => {
        let palette = new Palette(paletteLength, stops, ...refCanvasColors);
        expect(palette.colors).toMatchSnapshot();
        expect(palette.colorStops).toEqual(stops);

        colors = palette.colors;

        expect(() => validate(new Palette(paletteLength, Ease.linear, stops, ...refCanvasColors))).not.toThrow();
        expect(() => validate(new Palette(paletteLength, Ease.linear, ...refCanvasColors))).not.toThrow();
        expect(() => validate(new Palette(paletteLength, ...refCanvasColors))).not.toThrow();

        expect(() => validate(new Palette("", paletteLength, ...refCanvasColors))).not.toThrow();
        expect(() => validate(new Palette("invalid", paletteLength, ...refCanvasColors))).not.toThrow();

        expect(() => validate(Palette.getColors(paletteLength, refColors, stops, "rgb"))).not.toThrow();

        palette = new Palette("lrgb", paletteLength, stops, ...refCanvasColors);
        expect(palette.colors).toMatchSnapshot();
        expect(palette.colorStops).toEqual(stops);

        colors = palette.colors;

        expect(() => validate(new Palette("lrgb", paletteLength, Ease.linear, stops, ...refCanvasColors))).not.toThrow();
        expect(() => validate(new Palette("lrgb", paletteLength, Ease.linear, ...refCanvasColors))).not.toThrow();
        expect(() => validate(new Palette("lrgb", paletteLength, ...refCanvasColors))).not.toThrow();

        expect(() => validate(Palette.getColors(paletteLength, refColors, stops, "lrgb"))).not.toThrow();

        palette = new Palette("hsl", paletteLength, stops, ...refCanvasColors);
        expect(palette.colors).toMatchSnapshot();
        expect(palette.colorStops).toEqual(stops);

        colors = palette.colors;

        expect(() => validate(new Palette("hsl", paletteLength, Ease.linear, stops, ...refCanvasColors))).not.toThrow();
        expect(() => validate(new Palette("hsl", paletteLength, Ease.linear, ...refCanvasColors))).not.toThrow();
        expect(() => validate(new Palette("hsl", paletteLength, ...refCanvasColors))).not.toThrow();

        expect(() => validate(Palette.getColors(paletteLength, refColors, stops, "hsl"))).not.toThrow();
    });
});
