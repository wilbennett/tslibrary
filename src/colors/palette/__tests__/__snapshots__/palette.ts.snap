// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should create palettes Should allow creating Palette instance 1`] = `
Array [
  0,
  0.3333333333333333,
  0.6666666666666666,
  1,
]
`;

exports[`Should create palettes Should allow creating Palette instance 2`] = `
Array [
  SColor {
    "_converter": RgbConverter {},
    "_flags": 8,
    "_value": -16776961,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 1,
    "_value": -11184641,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 1,
    "_value": -5592321,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 8,
    "_value": -1,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 1,
    "_value": -1428903169,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 1,
    "_value": 1420449023,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 8,
    "_value": 8388863,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 1,
    "_value": 5592575,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 1,
    "_value": 2796031,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 8,
    "_value": 65535,
  },
]
`;

exports[`Should create palettes Should allow creating Palette instance 3`] = `
Array [
  SColor {
    "_converter": RgbConverter {},
    "_flags": 8,
    "_value": -16776961,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 1,
    "_value": -7105537,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 1,
    "_value": -3092225,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 8,
    "_value": -1,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 1,
    "_value": -790834945,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 1,
    "_value": -1816882177,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 8,
    "_value": 8388863,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 1,
    "_value": 6853631,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 1,
    "_value": 4837631,
  },
  SColor {
    "_converter": RgbConverter {},
    "_flags": 8,
    "_value": 65535,
  },
]
`;

exports[`Should create palettes Should allow creating Palette instance 4`] = `
Array [
  SColor {
    "_converter": HslConverter {},
    "_flags": 8,
    "_value": 1644900,
  },
  SColor {
    "_converter": HslConverter {},
    "_flags": 1,
    "_value": 1106404,
  },
  SColor {
    "_converter": HslConverter {},
    "_flags": 1,
    "_value": 551396,
  },
  SColor {
    "_converter": HslConverter {},
    "_flags": 8,
    "_value": 12900,
  },
  SColor {
    "_converter": HslConverter {},
    "_flags": 1,
    "_value": 84436452,
  },
  SColor {
    "_converter": HslConverter {},
    "_flags": 1,
    "_value": 168876388,
  },
  SColor {
    "_converter": HslConverter {},
    "_flags": 8,
    "_value": 253299940,
  },
  SColor {
    "_converter": HslConverter {},
    "_flags": 1,
    "_value": 337187044,
  },
  SColor {
    "_converter": HslConverter {},
    "_flags": 1,
    "_value": 421074276,
  },
  SColor {
    "_converter": HslConverter {},
    "_flags": 8,
    "_value": 504961380,
  },
]
`;
